### DISCUSSION TITLE
Failure handling when Matching optionals

### DISCUSSION THREAD

**Discussion by chichid** [pest-parser/pest#608] [Q&A] [answered]
Currently the following expression will always succeed for the B segments...
```
A = { "a" ~ B* }
B = { "b" }
```
How can we implement grammar so that the rule is optional, and it only fails when something is provided?

So for the above example:
- a would succeed
- abbbb would succeed
- but a##bb would fail

**Comment by CAD97** [comment, marked as answer]
The EOI rule requires the End of Input, so `{ "a" ~ B* ~ EOI }` works here as specifically what you asked for.

More generally, you need to use `&` or `!` lookahead; you could use `!"#"` to require that the following is not `#`, or `&"b"` to require that a `b` follows.

As an example:

```
WHITESPACE = _{ WHITE_SPACE }

list = { (a | error)* }

error = @{ (!WHITESPACE ~ ANY)+ }

a = @{ "a" ~ (b+ | &(WHITESPACE | EOI)) }
b = @{ "b" }
```

```
a abbbb a##bb a
```

```yaml
- list
  - a: "a"
  - a: "abbbb"
  - error: "a##bb"
  - a: "a"
```

**Reply by chichid** [reply]
thanks @CAD97. The error technic is very interesting stuff!
